flight/Modules/VtolPathFollower/vtol_follower_control.c

```diff
@@ -496,22 +389,69 @@ static int32_t vtol_follower_control_accel(float dT)
 	return 0;
 }
 
+
+static float vtol_follower_control_altitude(float downCommand) {
+	AltitudeHoldStateData altitudeHoldState;
+	altitudeHoldState.VelocityDesired = downCommand;
+	altitudeHoldState.Integral = vtol_pids[DOWN_VELOCITY].iAccumulator / 1000.0f;
+	altitudeHoldState.AngleGain = 1.0f;
+
+	if (altitudeHoldSettings.AttitudeComp > 0) {
+		// Throttle desired is at this point the mount desired in the up direction, we can
+		// account for the attitude if desired
+		AttitudeActualData attitudeActual;
+		AttitudeActualGet(&attitudeActual);
+
+		// Project a unit vector pointing up into the body frame and
+		// get the z component
+		float fraction = attitudeActual.q1 * attitudeActual.q1 -
+				 attitudeActual.q2 * attitudeActual.q2 -
+				 attitudeActual.q3 * attitudeActual.q3 +
+				 attitudeActual.q4 * attitudeActual.q4;
+
+		// Add ability to scale up the amount of compensation to achieve
+		// level forward flight
+		fraction = powf(fraction, (float) altitudeHoldSettings.AttitudeComp / 100.0f);
+
+		// Dividing by the fraction remaining in the vertical projection will
+		// attempt to compensate for tilt. This acts like the thrust is linear
+		// with the output which isn't really true. If the fraction is starting
+		// to go negative we are inverted and should shut off throttle
+		downCommand = (fraction > 0.1f) ? (downCommand / fraction) : 0.0f;
+
+		altitudeHoldState.AngleGain = 1.0f / fraction;
+	}
+
+	altitudeHoldState.Throttle = downCommand;
+	AltitudeHoldStateSet(&altitudeHoldState);
+
+	return downCommand;
+}
+
 /**
  * Compute desired attitude from the desired velocity
  * @param[in] dT the time since last evaluation
+ * @param[in] att_adj an adjustment to the attitude for loiter mode
  *
  * Takes in @ref NedActual which has the acceleration in the
  * NED frame as the feedback term and then compares the
  * @ref VelocityActual against the @ref VelocityDesired
  */
-int32_t vtol_follower_control_attitude(float dT)
+int32_t vtol_follower_control_attitude(float dT, const float *att_adj)
 {
 	vtol_follower_control_accel(dT);
 
+	float default_adj[2] = {0,0};
+
+	if (!att_adj) {
+		att_adj = default_adj;
+	}
+
 	AccelDesiredData accelDesired;
 	AccelDesiredGet(&accelDesired);
 
-	StabilizationDesiredData stabDesired;
+	StabilizationSettingsData stabSet;
+	StabilizationSettingsGet(&stabSet);
```

Comment:
Hm.  I'm not sure I buy this.  And if it is an overhead problem I should just fix the uavobject manager.

>  the most appropriate thing is the updated callback actually sets a flag. then at the top of the while loop we consume that flag and update the settings. this avoids any race conditions where a setting can change mix-execution with potentially unpredicted side effects.

Wait, I don't understand.  We take a copy of it while protecting accesses to the master with a lock, right?  So getting it at the top of the function doesn't risk it changing mid-execution...

In any event, this was a simple change in factoring-- I changed the later, pre-existing calls to  StabilizationSettingsManualRateGet(manual_rate); to instead look at a copy of stabSet retrieved at the beginning of the function, which if anything should be safer in these respects.  I'd rather not tangle up this pull with additional cleanup.